app/core/models/ocelot/stanzas/CalculationStanza.scala

```diff
@@ -59,25 +61,50 @@ sealed trait Operation {
 
   def eval(labels: Labels): Labels
 
-  def value(arg: String, labels: Labels): String = labelReference(arg).fold(arg) { ref =>
-    labels.value(ref).getOrElse("")
+  def value(arg: String, labels: Labels): Option[String] = labelReference(arg).fold[Option[String]](Some(arg)){ref =>
+    labels.value(ref)
+  }
+
+  def listValue(arg: String, labels: Labels): Option[List[String]] = {
+    labelReference(arg).fold[Option[List[String]]](None){ref => labels.valueAsList(ref)}
+  }
+
+  def operands(labels: Labels): (Option[String], Option[List[String]], Option[String], Option[List[String]]) = {
+
+    val xScalar: Option[String] = value(left, labels)
+    val yScalar: Option[String] = value(right, labels)
+
+    (xScalar, yScalar) match {
+      case (Some(_), Some(_)) => (xScalar, None, yScalar, None) // Optimize for scalar on scalar as these are the most common operations
+      case _ =>
+        val xList: Option[List[String]] = listValue(left, labels)
+        val yList: Option[List[String]] = listValue(right, labels)
+
+        (xScalar, xList, yScalar, yList)
+    }
   }
 
-  def unsupportedOperation(operationName: String)(arg1: Any, arg2: Any): Option[String] = {
+  def unsupportedOperation(operationName: String)(arg1: Any, arg2: Any ): Option[String] = {
```

Comment:
these two fns could be combined and parameterized as they always return None  
`def unsupportedOperation[A](operationName: String)(arg1: Any, arg2: Any ): Option[A]`